### PR TITLE
Use core's SDKProblem for errors

### PR DESCRIPTION
**Pull request by eiri**
## PR summary

- Add GetComponentInfo to support new Go core problem types.
- Move base service into package `base` to avoid circular dependencies
- Convert errors in our custom packages  to core's `SDKProblem`

Fixes: #468

**Note: An existing issue is [required](https://github.com/IBM/cloudant-go-sdk/blob/main/CONTRIBUTING.md#PRs) before opening a PR.**

## PR Checklist

Please make sure that your PR fulfills the following requirements:

- [x] The commit message follows the
[Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: auth/couchdb_session_authenticator.go [hunk 1]
```diff
@@ -221,8 +226,13 @@ func (a *CouchDbSessionAuthenticator) requestSession() (*session, error) {
 			Headers:    resp.Header,
 			RawResult:  buff.Bytes(),
 		}
-		err := fmt.Errorf(buff.String())
-		return nil, core.NewAuthenticationError(detailedResponse, err)
+
+		err := &core.HTTPProblem{
+			IBMProblem: core.IBMErrorf(nil, core.NewProblemComponent("", ""), buff.String(), ""),
```
Comment (ricellis) [inline auth/couchdb_session_authenticator.go:231]: a little unsure on this empty component

Comment (eiri) [reply auth/couchdb_session_authenticator.go:231]: Yes, this one is awkward. Core [deprecated `NewAuthenticationError`](https://github.com/IBM/go-sdk-core/blob/main/core/authentication_error.go#L28) which is a pity, it was one handy util function.

The recommended `HTTPProblem` doesn't have initializer, so I lifted [code from `httpErrorf`](https://github.com/IBM/go-sdk-core/blob/main/core/authentication_error.go#L48) and the empty component comes from [the function itself](https://github.com/IBM/go-sdk-core/blob/main/core/http_problem.go#L140)

It has `buff.String()` as a summary which is our error message, so it's not entirely empty.

Theoretically we can also run it through [`EnrichHTTPProblem`](https://github.com/IBM/go-sdk-core/blob/main/core/http_problem.go#L149), but then we'd have to define `operationID` and `ProblemComponent` for it at which moment I kind of gave up.

Comment (eiri) [reply auth/couchdb_session_authenticator.go]: I thought about it a little bit more and I think you right, we can do better than empty struct here, even if that's more verbose. Updated in https://github.com/IBM/cloudant-go-sdk/pull/474/commits/fa5d41a6dafe11c883893d7f5b2c0199703d8c1c

Comment (ricellis) [reply auth/couchdb_session_authenticator.go]: Thanks. I wonder if we should ask for a `HTTPProblem` initializer - I was trying to think if there were any use cases beyond auth that might make it more widely applicable.

File: auth/couchdb_session_authenticator.go [hunk 2]
```diff
@@ -185,6 +186,10 @@ func (a *CouchDbSessionAuthenticator) requestSession() (*session, error) {
 	builder, err := core.NewRequestBuilder(core.POST).
 		ResolveRequestURL(a.URL, "/_session", nil)
 	if err != nil {
+		// This is hopefully a temp override before core fixes ResolveRequestURL
+		if err.Error() == core.ERRORMSG_SERVICE_URL_MISSING {
+			err = core.SDKErrorf(nil, err.Error(), "empty-url", common.GetComponentInfo())
+		}
```
Comment (ricellis) [inline auth/couchdb_session_authenticator.go:192]: #476 upgraded core to `5.16.3` so hopefully this can revert.

**Review by ricellis — APPROVED** [review]
+1, new core is in, so should be able to remove the resolve url thing before merging

**Pull request merged**